libs/ledger/src/chain/main_chain.cpp

```diff
@@ -146,6 +145,88 @@ BlockStatus MainChain::AddBlock(Block const &blk)
   return status;
 }
 
+/**
+ * Internal: add a parent-child forward reference if it is unknown yet.
+ *
+ * @param parent
+ * @param child
+ */
+template <class ParentHash, class ChildHash>
+void MainChain::CacheReference(ParentHash &&parent, ChildHash &&child,
+                               IntBlockPtr parent_block) const
+{
+  auto siblings{references_.equal_range(parent)};
+  auto sibling_it{std::find_if(siblings.first, siblings.second,
+                               [&child](auto const &ref) { return ref.second == child; })};
+  if (sibling_it != siblings.second)
+  {
+    // this child has been already referred to
+    return;
+  }
+
+  switch (references_.count(parent))
+  {
+  case 0:  // this one will be an unique forward ref
+    if (parent_block || LookupBlockFromCache(parent, parent_block))
+    {
+      parent_block->next_hash = child;
+    }
+    break;
+  case 1:  // that forward ref from parent is no longer unique
+    if (parent_block || LookupBlockFromCache(parent, parent_block))
+    {
+      parent_block->next_hash = BlockHash{};
+    }
+    break;
+  default:;
+  }
+  references_.emplace_hint(siblings.first, std::forward<ParentHash>(parent),
+                           std::forward<ChildHash>(child));
+}
```

Comment:
This is hard to follow if you are jumping into this code. Could I suggest the following (and for the forget also):
```
  // Get {begin, end} iterator to all known siblings of this block
  auto siblings = forward_references_.equal_range(parent);

  auto &fwd_siblings_begin = siblings.first;
  auto &fwd_siblings_end = siblings.second;

  // Find any instances in this where the value is our child
  auto sibling_it = std::find_if(fwd_siblings_begin, fwd_siblings_end,
                               [&child](auto const &ref) { return ref.second == child; });

  if (sibling_it != fwd_siblings_end)
  {
    // this child has been already referred to
    return;
  }

  // count how many forward references are already known for our hash - note
  // 2 or more fills the block with an empty hash
  switch (references_.count(parent))
  {
  case 0:  // this one will be an unique forward ref
    if (parent_block || LookupBlockFromCache(parent, parent_block))
    {
      parent_block->next_hash = child;
    }
    break;
  case 1:  // that forward ref from parent is no longer unique
    if (parent_block || LookupBlockFromCache(parent, parent_block))
    {
      parent_block->next_hash = BlockHash{};
    }
    break;
  default:;
  }
  references_.emplace_hint(fwd_siblings_begin, std::forward<ParentHash>(parent),
                           std::forward<ChildHash>(child));
```
Stylistically, I think we prefer = to {} construction.

Reply:
> I think we prefer = to {} construction.

It's my personal idiosyncrasy: I am afraid to use copy initialization when copy elision is not standard yet. :-\

Reply:
Does this not do RVO which is standard, in this instance?

Reply:
Quite likely.